libzim/libzim.pyx

```diff
@@ -0,0 +1,286 @@
+cimport libzim
+cimport cpython.ref as cpy_ref
+from cython.operator import dereference
+
+from libc.stdint cimport uint32_t, uint64_t
+from libcpp.string cimport string
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr, make_shared
+
+import datetime
+
+from collections import defaultdict
+
+#########################
+#       ZimBlob         #
+#########################
+
+cdef class ZimBlob:
+    cdef Blob* c_blob
+
+    def __init__(self, content):
+
+        if isinstance(content, str):
+            ref_content = content.encode('UTF-8')
+        else:
+            ref_content = content
+
+        self.c_blob = new Blob(<char *> ref_content, len(ref_content))
```

Comment:
We must keep a reference to `ref_content`.
Else, it may be deleted by the gc and the blob will point to nothing.

Reply:
As far as I know when I pass to a parameter to a cython function it auto creates the reference when I do assignment
Please take a look here:
https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#python-objects-as-parameters-and-return-values

that's the reason I don't type it.

I also assign to ref_content so that this protects

https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#caveats-when-using-a-python-string-in-a-c-context

Reply:
The "auto reference increment" of cython works well when your function only use the parameter:
- increment ref to pyobject
- "wrap" pyobject into a cobject
- pass cobject to function
- use cobject in function
- quit function
- decrement ref to pyobject.

But now we are not directly using it. We are storing the cobject to use it later:
- increment ref to (ref_content) pyobject
- "wrap" pyobject into a cobject
- pass cobject to (c)Blob constructor
- store cobject in (c)Blob
- quit constructor
- decrement ref to (ref_content) pyobject.
- Store the (c)blob in (py)Blob
- quit (py)Blob.__init__ (and so left all reference to `ref_content`)
- [Later] Use Blob. <--- But ref_content doesn't exist.

I think we have to keep the `ref_content` in (py)Blob should be enough (so ref_content exist as long as (py)Blob exist).
And we sure that we keep a reference to (py)Blob when we pass the (c)Blob to the c world.
The best is probably to use a intermediate method in `Article` (it takes the (py)Blob return by the `get_data` and keep a ref in the article before returning it (where the blob will be converted to c in `blob_cy_call_fct`).
It will be good as long as libzim doesn't try to read the blob after the (py)Article has been delete.
Libzim ensure that it will keep a reference to the std::shared_ptr<zim::creator::Article>` as long as it need it (and so use blob).
So the wrapper should be sure that we keep a reference to the (py)Article as long as the `zim::creator::article` is not deleted. (But I think we are good here because the Article wrapper increment the ref on the python article).